### PR TITLE
Feature/initial implementation

### PR DESCRIPTION
**Pull request by sitepark-veltrup**
See
- https://sitepark.github.io/atoolo-docs/develop/bundles/form/
- https://sitepark.github.io/atoolo-docs/develop/form/

### CODE REVIEW

File: .github/workflows/verify.yml [hunk 1]
```diff
@@ -12,6 +12,8 @@ on:
 jobs:
   verify:
     uses: sitepark/github-project-workflow/.github/workflows/composer-verify.yml@release/1.x
+    with:
+      phpVersions: '["8.2","8.3"]'
```
Comment (sitepark-schaeper) [inline .github/workflows/verify.yml]: why do we verify with 8.2 and 8.3 but release with 8.4?

File: src/Dto/FormSubmission.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ * @codeCoverageIgnore
+ */
+class FormSubmission
+{
+    public bool $approved = false;
+
+    public function __construct(
+        public readonly string $remoteAddress,
+        public readonly FormDefinition $formDefinition,
+        public readonly object $data,
```
Comment (sitepark-schaeper) [inline src/Dto/FormSubmission.php]: why `object`? could we not narrow it down to atleast `stdClass`?

File: src/Processor/EmailSender.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+     *     cc?: array<array{address: string, name: string}>,
+     *     bcc?: array<array{address: string, name: string}>,
+     *     subject?: string,
+     *     format?: 'html'|'text',
+     *     showEmpty?: bool,
+     *     attachCsv?: bool,
+     * } $options
+     * @throws Html2TextException
+     * @throws TransportExceptionInterface
+     * @throws Exception
```
Comment (sitepark-schaeper) [inline src/Processor/EmailSender.php:41]: what `Exception`? under what circumstances?

File: src/Processor/EmailSender.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        private readonly EmailMessageModelFactory $modelFactory,
+        private readonly EmailHtmlMessageRenderer $htmlMessageRenderer,
+        private readonly CsvGenerator $csvGenerator,
+        private readonly MailerInterface $mailer,
+    ) {}
+
+    /**
+     * @param array{
+     *     from: array<array{address: string, name: string}>,
+     *     to: array<array{address: string, name: string}>,
```
Comment (sitepark-schaeper) [inline src/Processor/EmailSender.php]: I believe these two should be `non-empty-array`s as it would be senseless to send an email without either.

File: src/Processor/EmailSender.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+     *     cc?: array<array{address: string, name: string}>,
+     *     bcc?: array<array{address: string, name: string}>,
```
Comment (sitepark-schaeper) [inline src/Processor/EmailSender.php:34]: There is no need for these to be nullable, their values can just be empty `array`s.

File: src/Processor/EmailSender.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+     * @throws Html2TextException
+     * @throws TransportExceptionInterface
+     * @throws Exception
+     */
+    public function process(FormSubmission $submission, array $options): FormSubmission
+    {
+        $rendererModel = $this->modelFactory->create($submission, $options['showEmpty'] ?? false);
+        $result = $this->htmlMessageRenderer->render($rendererModel);
+        $html = $result->html;
+        $text = Html2Text::convert($html);
```
Comment (sitepark-schaeper) [inline src/Processor/EmailSender.php]: Would it be "better" to move the responsibility of providing an alternative email-message in text format into the `MessageRenderer`?

File: src/Processor/SubmitProcessor.php [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+use Atoolo\Form\Dto\FormSubmission;
+use Symfony\Component\DependencyInjection\Attribute\AutoconfigureTag;
+
+#[AutoconfigureTag('atoolo_form.processor')]
+interface SubmitProcessor
+{
+    /**
+     * @param array<string,mixed> $options
+     */
+    public function process(FormSubmission $submission, array $options): FormSubmission;
```
Comment (sitepark-schaeper) [inline src/Processor/SubmitProcessor.php:16]: I feel like it would be more practical to instead return some sort of `ProcessingResult` to indicate one of the following usecases I've seen in the implementations:
- have a submission skip all following validators (`Immune` or `Whitelisted`)
- block a submission (`Denied` or `Blocked` with a `Reason` to properly forumlate a response)
- forward a submission to the next processor (`Okay` or `Allowed`)

We would not need a mutable `approved` property in the `FormSubmission`, which we currently have to check in each processsor implementation separatly. as it currently stands we also have to create and throw exceptions whenever a processor needs to signal to stop processing.

File: src/Service/DataUrlParser.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        $parameterString = $matches[2];
+
+        $parameters = [];
+        $base64Data = '';
+
+        foreach (explode(';', $parameterString) as $parameter) {
+            if (str_starts_with($parameter, 'base64,')) {
+                $base64Data = substr($parameter, 7);
+            } else {
+                [$key, $value] = explode('=', $parameter, 2);
```
Comment (sitepark-schaeper) [inline src/Service/DataUrlParser.php:28]: The `$parameter` may not contain an equals sign, which would cause a PHP Warning with `$value` beeing set to `null` (which `urldecode` transforms into an empty string).
I belive we should try to catch this case.

File: src/Service/DataUrlParser.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+        $binaryData = base64_decode($base64Data, true);
+        if ($binaryData === false) {
+            throw new DataUrlException('invalid base64');
+        }
+
+        return new UploadFile(
+            filename: $parameters['name'] ?? '',
+            contentType: $mimeType,
+            data: $binaryData,
```
Comment (sitepark-schaeper) [inline src/Service/DataUrlParser.php:41]: I feel like the `$data` should not be an empty string if none was provided

File: src/Service/Email/CsvGenerator.php [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+use League\Csv\Writer;
+
+class CsvGenerator
+{
+    /**
+     * @param array{
+     *     items: ?array<EmailMessageModelItem>,
+     * } $model
+     * @throws CannotInsertRecord
+     * @throws Exception
```
Comment (sitepark-schaeper) [inline src/Service/Email/CsvGenerator.php:18]: `CannotInsertRecord` seems like a internal error that should be handled in this function.
`Exception` is a bit broad and lacks an explaination.

File: src/Service/Email/EmailHtmlMessageTwigMjmlRenderer.php [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+     * @throws RuntimeError
+     * @throws LoaderError
+     */
+    public function render(array $model): EmailHtmlMessageRendererResult
+    {
+        $mjml = $this->twig->render('@AtooloForm/email.mjml.html.twig', $model);
+        return new EmailHtmlMessageRendererResult(
+            subject: '',
+            html: $this->mjmlRenderer->render($mjml),
+            attachments: [],
```
Comment (sitepark-schaeper) [inline src/Service/Email/EmailHtmlMessageTwigMjmlRenderer.php]: Why can the `EmailHtmlMessageTwigRenderer` have attachments but not the `EmailHtmlMessageTwigMjmlRenderer`?

File: src/Service/Email/EmailHtmlMessageTwigMjmlRenderer.php [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+     * @param EmailMessageModel $model
+     * @throws SyntaxError
+     * @throws RuntimeError
+     * @throws LoaderError
+     */
+    public function render(array $model): EmailHtmlMessageRendererResult
+    {
+        $mjml = $this->twig->render('@AtooloForm/email.mjml.html.twig', $model);
+        return new EmailHtmlMessageRendererResult(
+            subject: '',
```
Comment (sitepark-schaeper) [inline src/Service/Email/EmailHtmlMessageTwigMjmlRenderer.php]: why can't the twig renderes have a `subject`? and also, why is it an empty string?

File: src/Service/Email/EmailMessageModelFactory.php [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+            $includeEmptyFields,
+        );
+        $dateTime = $this->platform->datetime();
+        return [
+            'lang' => $submission->formDefinition->lang,
+            'url' => 'https://' . $this->channel->serverName,
+            'tenant' => [ 'name' => $this->channel->tenant->name ],
+            'host' => $this->channel->serverName,
+            'date' => $dateTime->format('d.m.Y'),
+            'time' => $dateTime->format('H:i:s'),
```
Comment (sitepark-schaeper) [inline src/Service/Email/EmailMessageModelFactory.php]: why can't we just use the `\DateTime`? this way we always have to remeber which is in what format - either when writing the factory or when using it (or even the created class for that matter). And maybe we don't even want it in that format - I don't think we should predetmine that.

File: src/Service/Email/MjmlRenderer.php [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * @codeCoverageIgnore
+ */
+class MjmlRenderer
+{
+    public function render(string $html): string
+    {
+        return $html;
+    }
+}
```
Comment (sitepark-schaeper) [inline src/Service/Email/MjmlRenderer.php:18]: What is this good for?

File: src/Service/FormDataModelFactory.php [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+    public function create(FormDefinition $definition, array $data, bool $includeEmptyFields): array
+    {
+        $this->stack = [];
+        $this->items = [];
+        $this->includeEmptyFields = $includeEmptyFields;
+        $reader = new FormReader($definition, $data, $this);
+        $reader->read();
+
+        return $this->items;
+    }
```
Comment (sitepark-schaeper) [inline src/Service/FormDataModelFactory.php:42]: this implementation is confusing to me.
what happens if a user is unaware of it and attempts to run this code - are empty fields included?
```php
$factory = new FormDataModelFactory($urlParser);
$reader = new FormReader($definition, $data, $factory);
$factory->create($definition, $data, includeEmptyFields: true);
$reader->read();
```

maybe it would make things clearer if we renamed `FromReaderHandler` to something like `FormReadingAggregator` with an abstract method `getAggregate` or `getResult`, which would have a generic type of what it aggregates. then we would not need this `create` function shortcut and have users do this:
```php
$aggregator = new FormDataModelFactory($urlParser, includeEmptyFields: true);
$reader = new FormReader($definition, $data, $aggregator);
$items = $reader->read();
```

File: src/Dto/UISchema/Control.php [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+class Control extends Element
+{
+    /**
+     * @param array<string,mixed> $options
+     * @param array<string,mixed>|null $htmlLabel
+     */
+    public function __construct(
+        public ?string $scope = null,
+        public string|bool|null $label = null,
```
Comment (sitepark-schaeper) [inline src/Dto/UISchema/Control.php:17]: why can `$label` be a `bool`?

File: src/Dto/UISchema/Control.php [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+declare(strict_types=1);
+
+namespace Atoolo\Form\Dto\UISchema;
+
+class Control extends Element
+{
+    /**
+     * @param array<string,mixed> $options
+     * @param array<string,mixed>|null $htmlLabel
```
Comment (sitepark-schaeper) [inline src/Dto/UISchema/Control.php]: What can a `$htmlLabel` value look like? I don't understand:
- why it's an array
- why it's `mixed`
- why it can be null

File: src/Dto/FormDefinition.php [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * @codeCoverageIgnore
+ */
+class FormDefinition
+{
+    /**
+     * @param JsonSchema $schema
+     * @param array<string, mixed> $data Default values of form fields that should be prefilled
+     * @param array<string, string>|null $buttons
```
Comment (sitepark-schaeper) [inline src/Dto/FormDefinition.php]: what are the keys and values of `$buttons`?

File: src/Dto/FormDefinition.php [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+     * @param array<string, array{
+     *     headline: string,
+     *     text: string,
+     * }>|null $messages
+     * @param array<string, array<string,mixed>>|null $processors
```
Comment (sitepark-schaeper) [inline src/Dto/FormDefinition.php:25]: what are the keys in both array dimensions?

File: src/Service/FormDataModelFactory.php [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+        $this->items[] = $data;
+    }
+
+    /**
+     * @param Control $control
+     * @param JsonSchema $schema
+     * @param string $name
+     * @param mixed $value
+     * @return void
+     */
```
Comment (sitepark-schaeper) [inline src/Service/FormDataModelFactory.php]: This comment really doesn't add anything

File: src/Dto/UISchema/Control.php [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+     */
+    public function __construct(
+        public ?string $scope = null,
```
Comment (sitepark-schaeper) [inline src/Dto/UISchema/Control.php]: What does `$scope` reference here and what values could it take?

File: src/Processor/EmailSender.php [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+use Symfony\Component\Mime\Address;
+use Symfony\Component\Mime\Email;
+
+#[AsTaggedItem(index: 'email-sender', priority: 10)]
+class EmailSender implements SubmitProcessor
+{
+    public function __construct(
+        private readonly EmailMessageModelFactory $modelFactory,
+        private readonly EmailHtmlMessageRenderer $htmlMessageRenderer,
+        private readonly CsvGenerator $csvGenerator,
```
Comment (sitepark-schaeper) [inline src/Processor/EmailSender.php:25]: The `CsvGenerator` is only needed if `$option['attachCsv']` is set to `true` - is that something we can determine beforehand and maybe tag a `CsvGenerator` as some sort of `AttachmentGenerator` or don't if we do not need it?

File: src/Service/FormDataModelFactory.php [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+            return 'radio-buttons';
+        }
+
+        if ($type === 'string' && isset($schema['oneOf'])
+        ) {
+            return 'select';
+        }
+
+        return 'text';
+    }
```
Comment (sitepark-schaeper) [inline src/Service/FormDataModelFactory.php:198]: I don't know the values `$schema` can contain, but this implementation would return `'text'` for basically anything since there is no validation.

File: src/Service/FormReader.php [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+        } elseif ($element instanceof Layout) {
+            $this->readLayout($element);
+        }
+    }
+
+    private function readControl(Control $control): void
+    {
+        if ($control->scope === null) {
+            return;
+        }
```
Comment (sitepark-schaeper) [inline src/Service/FormReader.php]: What does it mean for a `Control` not to have a `$scope`? Why are we not processing these cases? Or should this just straight up not be possible?

File: src/Service/JsonSchemaValidator.php [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+
+        $type = $constraint->getType();
+        $name = $constraint->getName();
+
+        $formatResolver->registerCallable(
+            $type,
+            $name,
+            function ($data, $schema) use ($constraint) {
+                return $constraint->check($data, $schema);
+            },
```
Comment (sitepark-schaeper) [inline src/Service/JsonSchemaValidator.php:64]: this function does not need to bind `$this`, it should probably be `static`.

File: src/Service/JsonSchemaValidator/DataUrlConstraint.php [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    /**
+     * @throws CustomError
+     */
+    public function check(mixed $value, stdClass $schema): bool
+    {
+        if (!is_string($value)) {
+            throw new CustomError('Value is not a string');
+        }
```
Comment (sitepark-schaeper) [inline src/Service/JsonSchemaValidator/DataUrlConstraint.php:35]: shouldn't this be something that is cought by the `JsonSchemaValidator`? I mean we don't also check weather or not a field expecting a number actually is one right? so why do we need to check if a `data-url` is a `string`? or could this constraint technically be configured to _any_ field? if so, is that not validated either? does this have to err at runtime?

File: src/Service/Platform.php [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+
+    /**
+     * @param array<mixed>|array<string,mixed> $array
+     * @throws \JsonException
+     */
+    public function arrayToObjectRecursive(array $array): object
+    {
+        $json = json_encode($array, JSON_THROW_ON_ERROR);
+        return (object) json_decode($json, false, 512, JSON_THROW_ON_ERROR);
+    }
```
Comment (sitepark-schaeper) [inline src/Service/Platform.php:48]: These don't seem like `Platform` reliant functions; It looks a bit like this is beeing (ab)used as a `Util` class.  
Are these functions really that commonly required? if so - why?

**Review by sitepark-schaeper — COMMENTED** [review]
First of all - I had to finish this up since it's getting late and github unfortunatly does not save pending comments. So some methods might not have gotten the attention they deserve, but with the rather architectual concerns I address here I think it's not worth investing too much time in these smaller chunks before those are settled.  
With that out of the way, I wrote some (unstructured) thoughts down that I had while reading through your code and would like to discuss them with you:

I've ignored the translations so far, but is the `.translation-cach` directory intended to be checked in?

Why are `Layout`s `Element`s?

The names of email specific classes are a bit wordy in my opinon. For example, `Dto/Email/EmailHtmlMessageRendererResult` could just be `Dto/Email/RenderingResult`, right?

A few classes here include the term `Model`, which I don't quite understand what it means in this context. May it just be a `Dto`?

There are a lot of mutable properties in the `Dto`s which I did not find why. Do we really need to expose these (and if so, can we properly differentiate them)?

We might have gone a little bit overboard with the array typing here. It's really hard to review it and figure out what is a class, what is just an array structure and which of their properties are actually gurantieed to be typed as defined and which might not. As such I did "skip" some of the implementations (like `FormDataModelFactory`) because I did not have the time to comprehend them. I feel like this complexity and mental overhead is not neccessary and we could just type out most of these as classes.  
This also applies to all the `$options` (for which it is also difficult to figure out what exactly is beeing configured and from where). I get that it's hard to declare these but having it as just `array<string, mixed>` but checking for explicit values seems like a red flag.

Another reoccuring theme seem to be nullable `array` types. I don't think there is a reason for an `array` that is allowed to be empty to also be nullable. an alternative would be something like `?non-empty-array<string, string>` but that is even less intuitive. 

Then there are classes with comments saying "not yet implemented". I get that creating them helps development but do they need to be checked in?

Lastly a couple of class comments could be nice. For example the `Constraint` interface might profit from a small sentence explaining what it is, why it exists and what it's used for. To be clear - I don't want to be pedantic and don't think it's too big of a deal, I just think someone that does not happen to have the context could have a hard time figuring this out where all it takes are 20ish words.

File: .github/workflows/verify.yml (hunk 1 above)
Comment (sitepark-veltrup) [reply .github/workflows/verify.yml]: One dependency is not yet PHP 8.4 ready. I'll change this when it's solved, see: https://github.com/phpro/api-problem-bundle/pull/27

File: src/Dto/FormSubmission.php (hunk 2 above)
Comment (sitepark-veltrup) [reply src/Dto/FormSubmission.php]: I first had this with `stdClass`. But there was a problem and I changed it. I changed it again and now there are no more problems. I'll leave it on `stdClass` for now

File: src/Processor/EmailSender.php (hunk 3 above)
Comment (sitepark-veltrup) [reply src/Processor/EmailSender.php:41]: `League\Csv\Exception`
https://csv.thephpleague.com/9.0/connections/#exceptions

File: src/Processor/EmailSender.php (hunk 5 above)
Comment (sitepark-veltrup) [reply src/Processor/EmailSender.php:34]: The data comes from the aggregate, which does not generate an empty array, but does not aggregate the key at all.
With `cc?` and `bb?` it is meant that the key is optional.

File: src/Processor/EmailSender.php (hunk 6 above)
Comment (sitepark-veltrup) [reply src/Processor/EmailSender.php]: Yes, there should actually be a separate `textMessageRenderer` here. I have saved the implementation for the moment. But it should follow at a later date. I'll add a TODO at this point.

File: src/Service/Email/CsvGenerator.php (hunk 10 above)
Comment (sitepark-veltrup) [reply src/Service/Email/CsvGenerator.php:18]: see `use League\Csv\Exception;`

File: src/Service/Email/EmailHtmlMessageTwigMjmlRenderer.php (hunk 11 above)
Comment (sitepark-veltrup) [reply src/Service/Email/EmailHtmlMessageTwigMjmlRenderer.php]: fixed

File: src/Service/Email/EmailHtmlMessageTwigMjmlRenderer.php (hunk 12 above)
Comment (sitepark-veltrup) [reply src/Service/Email/EmailHtmlMessageTwigMjmlRenderer.php]: No idea, I removed it.

File: src/Service/Email/EmailMessageModelFactory.php (hunk 13 above)
Comment (sitepark-veltrup) [reply src/Service/Email/EmailMessageModelFactory.php]: I have changed it.

File: src/Service/Email/MjmlRenderer.php (hunk 14 above)
Comment (sitepark-veltrup) [reply src/Service/Email/MjmlRenderer.php:18]: The method parameter name is not correct, it should be `$mjml`.
The class is not yet implemented. It is still unclear how we want to make the MJML template engine available. I have added a TODO here.

File: src/Service/FormDataModelFactory.php (hunk 15 above)
Comment (sitepark-veltrup) [reply src/Service/FormDataModelFactory.php:42]: This is about displaying the unfilled fields in the HTML email. This is a setting that an editor can set for each form. Since `FormDataModelFactory` is a service that is created via DI, but is used for all forms, this cannot be defined in the `FormDataModelFactory`.

A user would only ever use the service and not create an instance of the `FormDataModelFactory` themselves.

```php
    public function __construct(
        private readonly FormDataModelFactory $formDataModelFactory,
    ) {}

    public function doSomething()
    {
        $this->formDataModelFactory->create($definition, $data, includeEmptyFields: true);
    }
```

File: src/Dto/UISchema/Control.php (hunk 16 above)
Comment (sitepark-veltrup) [reply src/Dto/UISchema/Control.php:17]: https://jsonforms.io/docs/uischema/controls#label-string--boolean

File: src/Dto/UISchema/Control.php (hunk 17 above)
Comment (sitepark-veltrup) [reply src/Dto/UISchema/Control.php]: This is necessary to process HTML labels that are defined via the form editor in Infosite. This is not a simple string but a data structure. You can find it in the Sitekit.

Depending on the input element, there may be a "simple" string label, an HTML label or no label at all.

File: src/Service/Email/EmailHtmlMessageTwigMjmlRenderer.php (hunk 12 above)
Comment (sitepark-veltrup) [reply src/Service/Email/EmailHtmlMessageTwigMjmlRenderer.php]: added

File: src/Dto/FormDefinition.php (hunk 18 above)
Comment (sitepark-veltrup) [reply src/Dto/FormDefinition.php]: added

File: src/Dto/FormDefinition.php (hunk 19 above)
Comment (sitepark-veltrup) [reply src/Dto/FormDefinition.php:25]: This is variable and cannot be defined more precisely.
The key of the array is the key of a potential processor. the array `array<string,mixed>` is relevant options for this processor. Processors can have very different options.

File: src/Service/FormDataModelFactory.php (hunk 20 above)
Comment (sitepark-veltrup) [reply src/Service/FormDataModelFactory.php]: I removed the comment, I think it was autogenerated.

File: src/Dto/UISchema/Control.php (hunk 21 above)
Comment (sitepark-veltrup) [reply src/Dto/UISchema/Control.php]: https://jsonforms.io/docs/uischema/controls#scope-string

I have add a comment.

Scope is a mandatory field, however, so it cannot be null.

File: src/Processor/EmailSender.php (hunk 22 above)
Comment (sitepark-veltrup) [reply src/Processor/EmailSender.php:25]: Is it so bad if I inject a service that I may not need, depending on the case? I think that happens relatively often. Maybe you could mark it as lazy. But I think that's only worthwhile if it's relatively expensive to create and that's not the case here.

File: src/Service/FormDataModelFactory.php (hunk 23 above)
Comment (sitepark-veltrup) [reply src/Service/FormDataModelFactory.php:198]: Yes, text is the fallback.

UISchema and Json schema are used to identify which UI component is to be used.

See https://sitepark.github.io/atoolo-docs/develop/form/controls/#json-schema

Without fallback, it would mean that new components can only be generated once the FEDS module has been updated. With fallback, it makes the system more stable during an update process.

File: src/Service/FormReader.php (hunk 24 above)
Comment (sitepark-veltrup) [reply src/Service/FormReader.php]: Yes, scope cannot be null.

https://jsonforms.io/docs/uischema/controls#scope-string

File: src/Service/JsonSchemaValidator.php (hunk 25 above)
Comment (sitepark-veltrup) [reply src/Service/JsonSchemaValidator.php:64]: Here I don't understand what you mean. Where is `$this` bind.

File: src/Service/JsonSchemaValidator/DataUrlConstraint.php (hunk 26 above)
Comment (sitepark-veltrup) [reply src/Service/JsonSchemaValidator/DataUrlConstraint.php:35]: It could be that the JsonSchemaValidator already checks this beforehand, since we have defined `String` as the type.
But since the method is passed a `mixed`, I have to check that anyway, so that phpstan is satisfied.

File: src/Service/Platform.php (hunk 27 above)
Comment (sitepark-veltrup) [reply src/Service/Platform.php:48]: The methods are only used once. I have pulled them out so that I can test them individually.
I don't want an `Util` class. What could I call them?

File: src/Dto/UISchema/Control.php (hunk 17 above)
Comment (sitepark-veltrup) [reply src/Dto/UISchema/Control.php]: Actually I should specify it like this:

```
php @param array{text: string}|null $htmlLabel
``` 

but then I get the error from the Symfony deserializer:

```
Symfony\Component\Serializer\Exception\NotNormalizableValueException: The type of the "htmlLabel" attribute for class "Atoolo\Form\Dto\UISchema\Control" must be one of "array{text: string}", "null" ("array" given).
```

**Comment by sitepark-veltrup** [comment]
> Why are Layouts Elements?

https://jsonforms.io/docs/uischema/layouts#elements

Layouts can be nested.

Layouts have an `elements` attribute in which further layouts or controls can be contained. Both are elements of the `elements` list.

File: src/Processor/EmailSender.php (hunk 5 above)
Comment (sitepark-schaeper) [reply src/Processor/EmailSender.php:34]: oh yeah right, I got it mixed up for `?array` ..
